rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,527 @@
+produced whether the migration was run once or multiple times. This property
+should hold even if new (up to date) writes occur in between migration runs
+which introduces the following restrictions:
+
+### 4.3.1.1 Restrictions
+
+1. All document transforms need to be deterministic, that is a document
+   transform will always return the same result for the same set of inputs. 
+2. It should always be possible to construct the exact set of inputs required
+   for (1) at any point during the migration process (before, during, after). 
+
+Although these restrictions require significant changes, it does not prevent
+known upcoming migrations such as [sharing saved-objects in multiple spaces](https://github.com/elastic/kibana/issues/27004) or [splitting a saved
+object into multiple child
+documents](https://github.com/elastic/kibana/issues/26602). To ensure that
+these migrations are idempotent, they will have to generate new saved object
+id's deterministically with e.g. UUIDv5.
+
+### 4.3.1.2 Algorithm
+
+1. On startup, find the index to which the `.kibana` alias points,
+   `.kibana_n`, and check if any documents have an outdated `migrationVersion`
+   indicating that a migration needs to be performed. 
+2. Create `.kibana_n+1`, if it already exists, ignore the error and continue.
+3. Migrate documents by reading from `.kibana_n` and creating documents in
+   `.kibana_n+1`. If a document already exists, don't overwrite it and ignore.
+4. Move the `.kibana` index to point to `.kibana_n+1` (ignore if it has
+   already been moved). 
+
+Together with the limitations this algorithm ensures that migrations are
+idempotent. If two nodes are started simultaneously, both of them will start
+writing into `.kibana_n+1` but because migrations are idempotent, it doesn’t
+matter which node’s writes win.
+
+> Note: This algorithm doesn't support mixed Kibana versions attempting to
+> perform a migration at the same time. E.g. if a v7.8.0 and a v7.9.0 node
+> attempts to simultaneously migrate a 7.6.0 index the result will be an
+> inconsistent state. This is an unlikely scenario in practise and therefore
+> an acceptable limitation. In 8.x we can ensure that only the latest node
+> completes the migration. 
+
+## 4.5 Changes in 8.0
+1. Migrations will be in-place and re-use the same index for minor and patch
+   upgrades. 
+2. Only allow breaking field mapping changes (which requires creating a new
+   index) in a major.
+3. All update operations should use optimistic concurrency control.
+
+### 4.5.1 Migration algorithm (8.0):
+
+1. On startup, find the index to which the `.kibana` alias points,
+   `.kibana_n`, and check if any documents have an outdated `migrationVersion`
+   indicating that a migration needs to be performed.
+2. If there are documents with newer `migrationVersion` numbers, exit Kibana
+   with a fatal error.
+3. If there are outdated documents, migrate these in batches:
```

Comment:
TODO: add a step to update the mappings of the index

When two kibana nodes try to update the index to different mappings it's possible that there's a race condition and only one of the node's updates will succeed. This can happen when two nodes are different Kibana versions or have different plugins enabled. This could lead to write failures in step (4), we should catch write failures due to incompatible mappings and fail the migration if this happens.

If there are mixed versions i.e. 8.1 and 8.2 it is possible for 8.1 to win the mappings race, but then fail the migration because it detects the presence of documents migrated to v8.2. Later the 8.2 node's migration could also fail because of incompatible mappings (from 8.1). The next time 8.1 starts up it won't try to update the mappings because it detects 8.2 documents so will fail immediately. This will allow 8.2 to successfully set the mappings and complete the migration.